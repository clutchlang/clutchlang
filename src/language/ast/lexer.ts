export * from './lexer/token';
export * from './lexer/tokenizer';
