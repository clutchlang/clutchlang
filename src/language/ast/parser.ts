import * as tokens from './token';

/**
 * Represents an unresolved node in the abstract syntax tree.
 */
export abstract class AstNode {
  public abstract accept<R, C>(
    visitor: AstVisitor<R, C>,
    context?: C
  ): R | undefined;

  /**
   * First token that represents this AST node.
   */
  public abstract get first(): tokens.Token;

  /**
   * Last token that represents this AST node.
   */
  public abstract get last(): tokens.Token;
}

/**
 * Represents a simple AST node that originates from a single token.
 */
export abstract class SimpleNode extends AstNode {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an operator node parsed from a single token.
 */
export class Operator extends SimpleNode {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitOperator(this, context);
  }
}

/**
 * Represents a marker type of AST node for all statements.
 */
export abstract class Statement extends AstNode {}

/**
 * Represents a marker type of AST node for all expressions.
 */
export abstract class Expression extends Statement {}

/**
 * Represents a simple AST node that is also an expression.
 */
export abstract class SimpleExpression extends Expression {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an expression that utilizes an operator.
 */
export abstract class OperatorExpression extends Expression {
  constructor(
    public readonly operator: Operator,
    public readonly target: Expression
  ) {
    super();
  }
}

/**
 * Represents an expression in the form of `<OP><EXPR>`.
 */
export class PrefixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPrefixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.operator.first;
  }

  public get last(): tokens.Token {
    return this.target.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP>`.
 */
export class PostfixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPostfixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.operator.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP><EXPR>`.
 */
export class BinaryExpression extends OperatorExpression {
  constructor(
    operator: Operator,
    public readonly left: Expression,
    public readonly right: Expression
  ) {
    super(operator, left);
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitBinaryExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.left.first;
  }

  public get last(): tokens.Token {
    return this.right.last;
  }
}

/**
 * Represents an expression in the form of `<TARGET>.<PROPERTY>`.
 */
export class PropertyExpression extends Expression {
  constructor(
    public readonly target: Expression,
    public readonly property: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPropertyExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.property.last;
  }
}

/**
 * Represents a list of arguments passed to a function call.
 */
export class ArgumentList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly args: Expression[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitArgumentList(this, context);
  }
}

/**
 * Represents an expression for a function call.
 *
 * May _optionally_ contain a @member target.
 */
export class CallExpression extends Expression {
  constructor(
    public readonly name: Identifier,
    public readonly args: ArgumentList,
    public readonly target?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitCallExpression(this, context);
  }

  public get first(): tokens.Token {
    return (this.target || this.name).first;
  }

  public get last(): tokens.Token {
    return this.args.last;
  }
}

export class ConditionalExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly condition: Expression,
    public readonly body: Expression | StatementBlock,
    public readonly elseCondition?: Expression,
    public readonly elseBody?: Expression | StatementBlock
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitConditionalExpression(this, context);
  }

  public get last(): tokens.Token {
    return (this.elseBody || this.body).last;
  }
}

export class GroupExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression: Expression,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitGroupExpression(this, context);
  }
}

/**
 * Represents an expression for a literal identifier (name).
 */
export class Identifier extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitIdentifier(this, context);
  }

  public get name(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal boolean.
 */
export class LiteralBoolean extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralBoolean(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal number.
 */
export class LiteralNumber extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralNumber(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal string.
 */
export class LiteralString extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralString(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents a variable declared anywhere in scope.
 */
export class VariableDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly type?: Identifier,
    public readonly value?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitVariableDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return (this.value || this.type || this.name).last;
  }
}

/**
 * Represents a list of parameters for a function or closure definition.
 */
export class ParameterList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly params: VariableDeclaration[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitParameterList(this, context);
  }
}

/**
 * Represents a series of statements.
 */
export class StatementBlock extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly statements: Statement[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitStatementBlock(this, context);
  }
}

/**
 * Represents a return statement.
 */
export class ReturnStatement extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitReturnStatement(this, context);
  }

  public get last(): tokens.Token {
    return this.expression ? this.expression.last : this.first;
  }
}

/**
 * Represents a function declared anywhere in scope.
 */
export class FunctionDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly params: ParameterList,
    public readonly body: Expression | StatementBlock,
    public readonly returnType?: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitFunctionDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return this.body.last;
  }
}

/**
 * Represents a type declared anywhere in scope.
 */
export class TypeDeclaration extends AstNode {
  constructor(
    public readonly modifiers: tokens.Token[],
    public readonly name: Identifier,
    public readonly members: Array<FunctionDeclaration | VariableDeclaration>,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitTypeDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.modifiers.length ? this.modifiers[0] : this.name.first;
  }
}

/**
 *
 */
export class ModuleRoot extends AstNode {
  constructor(
    public readonly declarations: Array<
      FunctionDeclaration | TypeDeclaration | VariableDeclaration
    >
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitModuleRoot(this, context);
  }

  public get first(): tokens.Token {
    return this.declarations[0].first;
  }

  public get last(): tokens.Token {
    return this.declarations[this.declarations.length - 1].last;
  }
}

export class AstParser {}

/**
 * An AST visitor that by default does nothing visiting an AST node.
 *
 * It is intended to be a superclass for classes that use the visitor pattern
 * primarily as a dispatch mechanism (and do not need to recursively need to
 * visit the whole structure) and only need to visit a small number of node
 * types.
 */
export class AstVisitor<R, C> {
  public visitArgumentList(_: ArgumentList, __?: C): R | undefined {
    return;
  }

  public visitBinaryExpression(_: BinaryExpression, __?: C): R | undefined {
    return;
  }

  public visitCallExpression(_: CallExpression, __?: C): R | undefined {
    return;
  }

  public visitConditionalExpression(
    _: ConditionalExpression,
    __?: C
  ): R | undefined {
    return;
  }

  public visitFunctionDeclaration(
    _: FunctionDeclaration,
    __?: C
  ): R | undefined {
    return;
  }

  public visitGroupExpression(_: GroupExpression, __?: C): R | undefined {
    return;
  }

  public visitIdentifier(_: Identifier, __?: C): R | undefined {
    return;
  }

  public visitLiteralBoolean(_: LiteralBoolean, __?: C): R | undefined {
    return;
  }

  public visitLiteralNumber(_: LiteralNumber, __?: C): R | undefined {
    return;
  }

  public visitLiteralString(_: LiteralString, __?: C): R | undefined {
    return;
  }

  public visitModuleRoot(_: ModuleRoot, __?: C): R | undefined {
    return;
  }

  public visitOperator(_: Operator, __?: C): R | undefined {
    return;
  }

  public visitParameterList(_: ParameterList, __?: C): R | undefined {
    return;
  }

  public visitPrefixExpression(_: PrefixExpression, __?: C): R | undefined {
    return;
  }

  public visitPostfixExpression(_: PostfixExpression, __?: C): R | undefined {
    return;
  }

  public visitPropertyExpression(_: PropertyExpression, __?: C): R | undefined {
    return;
  }

  public visitReturnStatement(_: ReturnStatement, __?: C): R | undefined {
    return;
  }

  public visitStatementBlock(_: StatementBlock, __?: C): R | undefined {
    return;
  }

  public visitTypeDeclaration(_: TypeDeclaration, __?: C): R | undefined {
    return;
  }

  public visitVariableDeclaration(
    _: VariableDeclaration,
    __?: C
  ): R | undefined {
    return;
  }
}
