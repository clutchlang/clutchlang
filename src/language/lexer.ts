import { ISourceSpan, StringScanner } from '../agnostic/scanner';
import {
  Characters,
  isDigit,
  isHexadecimal,
  isLetter,
  isWhiteSpace,
} from '../agnostic/strings';

function isIdentifier(character: number): boolean {
  return isIdentifierStart(character) || isDigit(character);
}

function isIdentifierStart(character: number): boolean {
  return isLetter(character) || character === Characters.$_;
}

/**
 * Tokenizes and returns @param program as a series of @see Token.
 */
export function tokenize(program: string, onError?: ErrorReporter): Token[] {
  return new ClutchLexer().scanProgram(new StringScanner(program), onError);
}

/**
 * Reports a @param message at the provided @param span.
 */
export type ErrorReporter = (span: ISourceSpan, error: string) => void;

export enum TokenKind {
  // Literals
  IDENTIFIER = 'Identifier',
  STRING = 'String',
  NUMBER = 'Number',

  // Keywords
  CLASS = 'class',
  ELSE = 'else',
  FALSE = 'false',
  FOR = 'for',
  IF = 'if',
  LET = 'let',
  RETURN = 'return',
  SUPER = 'super',
  THEN = 'then',
  THIS = 'this',
  TRUE = 'true',
  WHILE = 'while',

  // Symbols
  ARROW = '->',
  LEFT_PAREN = '(',
  RIGHT_PAREN = ')',
  LEFT_CURLY = '{',
  RIGHT_CURLY = '}',

  // Operators
  PERIOD = '.',
  PLUS = '+',
  PLUS_EQUALS = '+=',
  MINUS = '-',
  MINUS_EQUALS = '-=',
  STAR = '*',
  STAR_EQUALS = '*=',
  SLASH = '/',
  SLASH_EQUALS = '/=',
  MODULUS = '%',
  MODULUS_EQUALS = '%=',
  EQUALS = '=',
  EQUALS_EQUALS = '==',
  EXCLAIM_EQUALS = '!=',
  EQUALS_EQUALS_EQUALS = '===',
  EXCLAIM_EQUALS_EQUALS = '!==',
  RIGHT_ANGLE = '>',
  RIGHT_ANGLE_EQUALS = '>=',
  LEFT_ANGLE = '<',
  LEFT_ANGLE_EQUALS = '<=',
  PIPE = '|',
  PIPE_PIPE = '||',
  AND = '&',
  AND_AND = '&&',
  CARET = '^',
  EXCLAIM = '!',
  TILDE = '~',
  PLUS_PLUS = '++',
  MINUS_MINUS = '--',
  LEFT_ANGLE_LEFT_ANGLE = '<<',
  RIGHT_ANGLE_RIGHT_ANGLE = '>>',

  // Misc
  EOF = '<EOF>',
}

/**
 * A language-level lexer that yields tokens understood by the grammar.
 */
export class ClutchLexer {
  public static readonly keywords: {
    [index: string]: TokenKind;
  } = {
    class: TokenKind.CLASS,
    else: TokenKind.ELSE,
    false: TokenKind.FALSE,
    for: TokenKind.FOR,
    if: TokenKind.IF,
    let: TokenKind.LET,
    return: TokenKind.RETURN,
    super: TokenKind.SUPER,
    this: TokenKind.THIS,
    true: TokenKind.TRUE,
    while: TokenKind.WHILE,
  };

  private program!: StringScanner;
  private onError!: ErrorReporter;
  private position!: number;
  private lastComments: IComment[] = [];

  public scanProgram(
    program: StringScanner,
    onError: ErrorReporter = (span, message) => {
      throw new SyntaxError(
        `${message} "${span.text}" at ${span.line}:${span.column}`
      );
    }
  ): Token[] {
    this.position = 0;
    this.program = program;
    this.onError = onError;
    const tokens: Token[] = [];
    while (program.hasNext()) {
      const token = this.scanToken();
      if (token) {
        tokens.push(token);
      }
    }
    tokens.push(this.createToken(TokenKind.EOF, ''));
    return tokens;
  }

  private scanToken(): Token | undefined {
    const character = this.program.advance();
    switch (character) {
      case Characters.$LPAREN: // "("
        return this.createToken(TokenKind.LEFT_PAREN);
      case Characters.$RPAREN: // ")"
        return this.createToken(TokenKind.RIGHT_PAREN);
      case Characters.$LCURLY: // "{"
        return this.createToken(TokenKind.LEFT_CURLY);
      case Characters.$RCURLY: // "}"
        return this.createToken(TokenKind.RIGHT_CURLY);
      case Characters.$PERIOD: // "."
        return this.createToken(TokenKind.PERIOD);
      case Characters.$PLUS: // "+" or "+=" or "++"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.PLUS_EQUALS
            : this.program.match(Characters.$PLUS)
              ? TokenKind.PLUS_PLUS
              : TokenKind.PLUS
        );
      case Characters.$MINUS: // "-" or "--" or "-=" or "->"
        return this.createToken(
          this.program.match(Characters.$RANGLE)
            ? TokenKind.ARROW
            : this.program.match(Characters.$EQUALS)
              ? TokenKind.MINUS_EQUALS
              : this.program.match(Characters.$MINUS)
                ? TokenKind.MINUS_MINUS
                : TokenKind.MINUS
          // Code coverage considers the `);` to be a missed line :(
          /* istanbul ignore next */
        );
      case Characters.$STAR: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.STAR_EQUALS
            : TokenKind.STAR
        );
      case Characters.$PERCENT: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.MODULUS_EQUALS
            : TokenKind.MODULUS
        );
      case Characters.$EQUALS: // "=" or "==" or "==="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EQUALS_EQUALS_EQUALS
              : TokenKind.EQUALS_EQUALS
            : TokenKind.EQUALS
        );
      case Characters.$LANGLE: // "<" or "<=" or "<<"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.LEFT_ANGLE_EQUALS
            : this.program.match(Characters.$LANGLE)
              ? TokenKind.LEFT_ANGLE_LEFT_ANGLE
              : TokenKind.LEFT_ANGLE
        );
      case Characters.$RANGLE: // ">" or ">=" or ">>"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.RIGHT_ANGLE_EQUALS
            : this.program.match(Characters.$RANGLE)
              ? TokenKind.RIGHT_ANGLE_RIGHT_ANGLE
              : TokenKind.RIGHT_ANGLE
        );
      case Characters.$SLASH: // "/" or "//" or "/="
        return this.scanSlash();
      case Characters.$SQUOTE: // "'"
        return this.scanString();
      case Characters.$PIPE: // "|" or "||"
        return this.createToken(
          this.program.match(Characters.$PIPE)
            ? TokenKind.PIPE_PIPE
            : TokenKind.PIPE
        );
      case Characters.$AND: // "&" or "&&"
        return this.createToken(
          this.program.match(Characters.$AND)
            ? TokenKind.AND_AND
            : TokenKind.AND
        );
      case Characters.$TILDE: // "~"
        return this.createToken(TokenKind.TILDE);
      case Characters.$CARET: // "^"
        return this.createToken(TokenKind.CARET);
      case Characters.$EXCLAIM: // "!" or "!=" or "!==""
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EXCLAIM_EQUALS_EQUALS
              : TokenKind.EXCLAIM_EQUALS
            : TokenKind.EXCLAIM
        );
      default:
        if (isWhiteSpace(character)) {
          this.position = this.program.position;
          break;
        }
        if (isDigit(character)) {
          return this.scanNumber(character);
        }
        if (isIdentifierStart(character)) {
          return this.scanIdentifierOrKeyword();
        }
        this.reportError(
          `Unexpected character "${String.fromCharCode(character)}"`
        );
    }
    return;
  }

  /**
   * Scans a string, returning `undefined` if there was no string.
   */
  private scanString(): Token {
    while (this.program.hasNext()) {
      if (this.program.match(Characters.$SQUOTE)) {
        this.position++;
        let lexeme = this.substring();
        lexeme = lexeme.substring(0, lexeme.length - 1);
        return this.createToken(TokenKind.STRING, lexeme);
      }
      this.program.advance();
    }
    this.reportError('Unterminated string');
    this.position++;
    return this.createToken(TokenKind.STRING);
  }

  /**
   * Scans any tokens that are valid numeric literals
   */
  private scanNumber(starting: number): Token {
    if (starting === Characters.$0) {
      if (
        this.program.match(Characters.$X) ||
        this.program.match(Characters.$x)
      ) {
        return this.scanHexNumber();
      }
    }
    if (
      this.program.match(Characters.$E) ||
      this.program.match(Characters.$e)
    ) {
      return this.scanDigits();
    }
    while (this.program.hasNext()) {
      const next = this.program.peek();
      if (this.program.match(Characters.$PERIOD)) {
        break;
      }
      if (!isDigit(next)) {
        break;
      }
      this.program.advance();
    }
    return this.scanDigits();
  }

  /**
   * Scans any tokens that are valid hexadecimel digits, returning a number.
   */
  private scanHexNumber(): Token {
    this.scanPredicate(isHexadecimal);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid digits, returning a number.
   */
  private scanDigits(): Token {
    this.scanPredicate(isDigit);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid identifiers.
   */
  private scanIdentifierOrKeyword(): Token {
    this.scanPredicate(isIdentifier);
    const identifierOrKeyword = this.substring();
    return new Token(
      ClutchLexer.keywords[identifierOrKeyword] || TokenKind.IDENTIFIER,
      identifierOrKeyword,
      this.clearComments(),
      this.position - identifierOrKeyword.length
    );
  }

  /**
   * Scans tokens preceding with "/".
   */
  private scanSlash(): Token | undefined {
    // "//"
    if (this.program.match(Characters.$SLASH)) {
      this.advanceToEOL();
      this.lastComments.push({
        lexeme: this.substring().trim(),
        offset: this.position,
      });
      return;
    }
    // "/" or "/="
    return this.createToken(
      this.program.match(Characters.$EQUALS)
        ? TokenKind.SLASH_EQUALS
        : TokenKind.SLASH
    );
  }

  /**
   * Scans until the EOF or until @param predicate returns false.
   */
  private scanPredicate(predicate: (character: number) => boolean): void {
    while (this.program.hasNext()) {
      const peek = this.program.peek();
      if (!predicate(peek)) {
        break;
      }
      this.program.advance();
    }
  }

  /**
   * Advances the position of the scanner until either EOF or EOL.
   *
   * **NOTE**: Treats `CR+LF` as a DOS-stlye line ending, not two lines.
   */
  private advanceToEOL(): void {
    while (this.program.hasNext()) {
      // "\n" (LF)
      if (this.program.match(Characters.$LF)) {
        return;
      }
      // "\r" (CR)
      if (this.program.match(Characters.$CR)) {
        // "\r\n" (CR+LF)
        this.program.match(Characters.$LF);
        return;
      }
      this.program.advance();
    }
  }

  /**
   * Returns a new token with a @param kind and the current substring.
   */
  private createToken(kind: TokenKind, content = this.substring()): Token {
    return new Token(
      kind,
      content,
      this.clearComments(),
      this.position - content.length
    );
  }

  /**
   * Clear and retrieve comments lexed.
   */
  private clearComments(): IComment[] {
    const comments = this.lastComments;
    this.lastComments = [];
    return comments;
  }

  /**
   * Returns the current substring, advancing the position counter.
   */
  private substring(): string {
    const start = this.position;
    const end = this.program.position;
    this.position = end;
    return this.program.substring(start, end);
  }

  /**
   * Reports a lexical @param error occurring at @param offset.
   */
  private reportError(error: string, offset = this.position): void {
    this.onError(this.program.source.span(offset, offset + 1), error);
  }
}

/**
 * A scanned token from @function tokenize.
 */
export interface IToken {
  readonly kind: TokenKind;
  readonly lexeme: string;
  readonly comments: IComment[];
  readonly offset: number;
}

/**
 * Implementation of @interface IToken private to @function tokenize.
 */
class Token implements IToken {
  constructor(
    public readonly kind: TokenKind,
    public readonly lexeme: string,
    public readonly comments: IComment[],
    public readonly offset: number
  ) {}
}

export interface IComment {
  readonly lexeme: string;
  readonly offset: number;
}
