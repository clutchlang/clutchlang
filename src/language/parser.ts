export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
