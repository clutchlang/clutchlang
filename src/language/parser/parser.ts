import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
