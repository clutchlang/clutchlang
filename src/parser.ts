export { Codes } from './parser/codes';
export { Lexer } from './parser/lexer';
export { Token } from './parser/token';
export { Tokens } from './parser/tokens';
