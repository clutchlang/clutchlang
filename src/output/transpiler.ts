import {
  AstCompilationUnit,
  AstFunctionDeclaration,
  AstInvocationExpression,
  AstLiteralBoolean,
  AstLiteralIdentifier,
  AstLiteralNumber,
  AstLiteralString,
  AstParenthesizedExpression,
  AstVariableDeclaration,
} from '../parser/source/ast/node';
import { AstVisitor } from '../parser/source/ast/visitor';

/**
 * A basic "compiler" that does a literal translation to JavaScript.
 *
 * It does *not*:
 * - Implement the Clutch language specification.
 * - Perform validation, type-checking, or any other error checking.
 *
 * This is a stop-gap output type until the pipeline is more complete.
 */
export class JsOutputTranspiler extends AstVisitor {
  public visitCompilationUnit(node: AstCompilationUnit): string {
    return node.functions.map(e => e.visit(this)).join('\n');
  }

  public visitFunctionDeclaration(node: AstFunctionDeclaration): string {
    return `
      function ${node.name}(${node.parameters.map(e => e.name).join(', ')}) {
        ${node.body.map(e => `${e.visit(this)};`).join('\n')}
      }
      ${node.name === 'main' ? 'main();' : ''}
    `;
  }

  public visitVariableDeclaration(node: AstVariableDeclaration): string {
    return `let ${node.name} = ${node.value.visit(this)}`;
  }

  public visitLiteralBoolean(node: AstLiteralBoolean): string {
    return node.value.toString();
  }

  public visitLiteralNumber(node: AstLiteralNumber): string {
    return node.value.toString();
  }

  public visitLiteralString(node: AstLiteralString): string {
    return `'${node.value.toString()}'`;
  }

  public visitLiteralIdentifier(node: AstLiteralIdentifier): string {
    return node.name.toString();
  }

  public visitInvocationExpression(node: AstInvocationExpression): string {
    const target = node.target;
    // Special cased for now.
    if (target instanceof AstLiteralIdentifier && target.name === 'print') {
      return `console.log(${node.args[0].visit(this)})`;
    }
    return `${target.visit(this)}(${node.args
      .map(e => e.visit(this))
      .join(', ')})`;
  }

  public visitParenthesizedExpression(
    node: AstParenthesizedExpression
  ): string {
    return `(${node.body.map(e => e.visit(this)).join(', ')})`;
  }
}
