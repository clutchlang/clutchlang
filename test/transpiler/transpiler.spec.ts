import { tokenize } from '../../src/language/lexer';
import { ClutchParser, FileRoot } from '../../src/language/parser';
import { SimpleJsTranspiler } from '../../src/transpiler/transpiler';

function parse(program: string): FileRoot {
  const tokens = tokenize(program);
  return new ClutchParser(tokens).parseFileRoot();
}

describe('JsOutputTranspiler', () => {
  it('should support "Hello World"', () => {
    const program = parse(`
      main -> {
        print('Hello World')
      }
    `);
    const visitor = new SimpleJsTranspiler();
    expect(program.accept(visitor).toString()).toMatchSnapshot();
  });

  it('should support most ast nodes', () => {
    const program = parse(`
      example -> {
        true
        1
        'Hello'
        foobar
        method(false)
        f(1 2)
        -1
        +1
        !true
        --1
        1++
        true && false
        false || true
      }
      explicit -> {
        return true
      }
      implicit -> {
        true
      }
      noReturn -> {
        let variable = 5
      }
      withIf -> if true then 1
      withIfElse -> if true then 1 else 2
      withIfStatement -> {
        if true then 
          print(1) 
        else
          print(2)
      }
      ternary(a b c) -> if a then b else c
    `);
    const visitor = new SimpleJsTranspiler();
    expect(program.accept(visitor).toString()).toMatchSnapshot();
  });

  it('should support parenthesized ast nodes', () => {
    const program = parse(`
      example -> {
        (true)
      }
    `);
    const visitor = new SimpleJsTranspiler();
    expect(program.accept(visitor).toString()).toMatchSnapshot();
  });

  it('should support functions with parameters', () => {
    const program = parse(`
      print2(a: String b: String): void -> {
        print(a)
        print(b)
      }
    `);
    const visitor = new SimpleJsTranspiler();
    expect(program.accept(visitor).toString()).toMatchSnapshot();
  });
});
