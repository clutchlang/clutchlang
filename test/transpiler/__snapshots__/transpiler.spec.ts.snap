// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsOutputTranspiler should support "Hello World" 1`] = `
"function main() {
  return  print  (  '  Hello World  '  )  ;
}
"
`;

exports[`JsOutputTranspiler should support functions with parameters 1`] = `
"function print2(a, b) {
  print  (  a  )  ;
  return  print  (  b  )  ;
}
"
`;

exports[`JsOutputTranspiler should support most ast nodes 1`] = `
"function example() {
  true  ;
  1  ;
  '  Hello  '  ;
  foobar  ;
  method  (  false  )  ;
  return  f  (  1  ,   2  )  ;
}
function explicit() {
  return  true  ;
}
function implicit() {
  true  ;
}
function noReturn() {
  let  variable  =  5  ;
}
function withIf() {
  true  ?  1  ;
}
function withIfElse() {
  true  ?  1  :  2  ;
}
function withIfStatement() {
  return  true  ?  print  (  1  )  :  print  (  2  )  ;
}
function ternary(a, b, c) {
  a  ?  b  :  c  ;
}
"
`;

exports[`JsOutputTranspiler should support parenthesized ast nodes 1`] = `
"function example() {
  return  (  true  )  (  false  )  ;
}
"
`;
